[PR/Fixes] Agent Service / Dual Stack Socket

markmandel opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/googleforgames/quilkin/blob/main/CONTRIBUTING.md 
   and developer guide https://github.com/googleforgames/quilkin/blob/main/build/README.md
2. Please label this pull request according to what type of issue you are addressing.
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/googleforgames/quilkin/blob/main/build/README.md#run-tests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup

> /kind documentation
> /kind feature
> /kind hotfix

**What this PR does / Why we need it**:

This fixes #744 so it passes tests, and does some other small cleanup. 

This includes:

* Dual stack UdpSocket
* Update Agent code to use Dual Stack Socket
* Fix shutdown channel in Agent test
* Upgrade to 50ms delay on ping test (sometimes failed for me after compilation at 10ms)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Closes #<issue number>`, or `Closes (paste link of issue)`.
-->

Also foundational work on #666.

**Special notes for your reviewer**:

Once this is merged, I can move on to using the DualStackLocalSocket throughout Quilkin, so we can finish #666.

Also, I rebased against main, so it created some mess (but I also fixed up the conflicts in #744) - apologies for that.

markmandel commented on src/utils/net.rs:
```diff
@@ -40,9 +52,69 @@ fn enable_reuse(sock: &Socket) -> io::Result<()> {
+    }
+
+    pub async fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
+        let mut v4_buf = vec![0; buf.len()];
```
Is there a better way than creating a whole new buffer each time? I can't reuse the same buffer in both places as Rust won't let you borrow something mutable more than once.

XAMPPRocky replied:
I think a reasonable first implementation would be to have two member buffers, and have the API return the slice (returning a slice should work here, but you might need to copy into a vec). This means slightly more initial memory, but will reduce the number of allocations.

markmandel replied:
I like it. If I need to copy to a vec, I might also look at returning an enum on which buffer was populated, so we don't have to create a whole vec each time.

markmandel commented:
Looks like https://developers.google.com/protocol-buffers now redirects to https://protobuf.dev/ - I'll file a separate PR  to fix.

markmandel commented on agones/src/xds.rs:
```diff
@@ -120,7 +120,10 @@ filters:
-            .insert(token_key.into(), base64::encode(token));
+            .insert(
+                token_key.into(),
+                base64::Engine::encode(&base64::engine::general_purpose::STANDARD, token),
```
Couldn't find it -- but this is identical to `main` - just was after the update to all the dependencies.

markmandel replied:
Oooh I see why this is the way it is. The function in utils is only `pub(crate)`, so it's not available to this test. I can update it to be `pub` in this PR?

XAMPPRocky replied:
Ah right I forgot this is just for the integration tests, this is fine then, we shouldn't expose the utils module

markmandel commented on src/protocol.rs:
```diff
@@ -32,65 +33,62 @@ const DISCRIMINANT_LEN: usize = 1;
+                match socket.recv_from(&mut v4_buf, &mut v6_buf).await {
+                    Ok((size, source)) => {
+                        let received_at = chrono::Utc::now().timestamp_nanos();
+                        let contents = match source {
```
Since we get a `SocketAddr` from `recv_from`, that actually ends up being a good indicator of which buffer to look at.

markmandel replied:
This felt like a good approach to me, as then we don't have to create a vec on each packet either.

XAMPPRocky replied:
Yeah I like this approach a lot 👍

markmandel replied:
Sweet!

Anything I can do to make this more easily mergable?

I'm happy to update `ep/agent-service` to `main` and resolve conflicts so that this is more easily merged, if you are swamped with other things.

XAMPPRocky replied:
Yeah I'm swamped right now, feel free to just push your commit to the PR

markmandel commented:
@XAMPPRocky I cleaned up the commits on both PRs, rebased #744 against `main`, and rebased this one as well.

Assuming it passes CI, this should be good to approve and merge into your PR, and then we can approve your PR, and everything should bubble up nicely.
